### PR TITLE
Move selected state to model and select file names

### LINKED ISSUE #132: The `selected` state should be intrinsic to the model.
Currently, the `selected` state on the model is used to pass view-specific state of the listing to other parts of the browser. It is set by the view listing on each update and is uni-directional. External changes to the `selected` state will not be reflected by the listing.

Instead, the model should be the source of truth for the selected state, and the listing widget should just render whatever the model says is selected. This will allow anything to change the selected state and will ensure that the selection does not get out of sync with the sort order.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #132.

### CODE REVIEW

File: src/listing.ts [hunk 1]
```diff
@@ -952,75 +955,78 @@ class DirListing extends Widget {
...
-
-      // Add the selected class to current row, and remove from all others.
-      for (let node of nodes) {
-        node.classList.remove(SELECTED_CLASS);
-      }
-      target.classList.add(SELECTED_CLASS);
+      selected = [name];
     }
+    this._model.selectedNames = selected;
+    this.update();
```
Comment (blink1073) [inline src/listing.ts:993]: I added this to force the update because the signal did not always fire.

**Comment by blink1073** [comment]
Notes to self:
- Expose `isSelected`, `select`, and `deselect` and fire a signal on those
- Fix cut class handling
- Make sure we aren't reading DOM classes outside of an event path

File: src/listing.ts [hunk 2]
```diff
@@ -557,6 +571,9 @@ class DirListing extends Widget {
     let content = utils.findElement(this.node, LIST_AREA_CLASS);
     let subtype = this.constructor as typeof DirListing;
 
+    this.node.classList.remove(MULTI_SELECTED_CLASS);
```
Comment (sccolbert) [inline src/listing.ts]: `this.removeClass(...)`

(only need to use `classList` for raw nodes)

File: src/listing.ts [hunk 3]
```diff
@@ -570,37 +587,25 @@ class DirListing extends Widget {
...
         }
       }
     }
 
-    // Update the selected items
-    this._updateSelected();
+    // Handle the selectors on the widget node.
+    let selectedNames = this._model.getSelected();
+    if (selectedNames.length > 1) {
+      this.node.classList.add(MULTI_SELECTED_CLASS);
```
Comment (sccolbert) [inline src/listing.ts]: `this.addClass`

File: src/model.ts [hunk 4]
```diff
@@ -132,6 +122,65 @@ class FileBrowserModel implements IDisposable {
   }
 
   /**
+   * Select an item by name.
+   *
+   * #### Notes
+   * This is a no-op if the name is not valid or already selected.
+   */
+  select(name: string): void {
+    if (name in this._selection && !this._selection[name]) {
```
Comment (sccolbert) [inline src/model.ts]: the `in` test is redundant, since `undefined` is falsey.

File: src/model.ts [hunk 5]
```diff
@@ -132,6 +122,65 @@ class FileBrowserModel implements IDisposable {
...
+   * Returns `true` for a valid name that is selected, `false` otherwise.
+   */
+  isSelected(name: string): boolean {
+    return !!this._selection[name];
+  }
+
+  /**
+   * Get the list of selected names.
+   */
+  getSelected(): string[] {
```
Comment (sccolbert) [inline src/model.ts:162]: If you like one-liners:
`Object.keys(this._selection).filter(k => this._selection[k]);`

File: src/model.ts [hunk 6]
```diff
@@ -132,6 +122,65 @@ class FileBrowserModel implements IDisposable {
...
+        output.push(key);
+      }
+    }
+    return output;
+  }
+
+  /**
+   * Clear the selected items.
+   */
+  clearSelected(): void {
```
Comment (sccolbert) [inline src/model.ts:169]: `this._selection = Object.create(null);`

File: src/model.ts [hunk 7]
```diff
@@ -38,6 +34,7 @@ class FileBrowserModel implements IDisposable {
   constructor(contentsManager: IContentsManager, sessionManager: INotebookSessionManager) {
     this._contentsManager = contentsManager;
     this._sessionManager = sessionManager;
+    this._selection = {};
```
Comment (sccolbert) [inline src/model.ts]: this._selection = Object.create(null);

File: src/model.ts [hunk 8]
```diff
@@ -424,14 +489,13 @@ class FileBrowserModel implements IDisposable {
...
 
-
   private _max_upload_size_mb = 15;
-  private _selectedIndices: number[] = [];
   private _contentsManager: IContentsManager = null;
   private _sessionIds: ISessionId[] = [];
   private _sessionManager: INotebookSessionManager = null;
   private _model: IContentsModel = null;
   private _kernelSpecs: IKernelSpecId[] = [];
+  private _selection: { [key: string]: boolean; } = null;
```
Comment (sccolbert) [inline src/model.ts]: If I initialize something in the constructor, I explicitly _do not_ initialize it inline as well. That way when I'm reading the code, I can look at see "oh, this must be initialized in the ctor"

File: src/model.ts (hunk 6 above)
Comment (blink1073) [inline src/model.ts:169]: Can't do that, I rely on all of the names being in there, how about this?

`Object.keys(this._selection).map(k => this._selection[k] = false);`

File: src/model.ts [hunk 9]
```diff
@@ -181,10 +230,26 @@ class FileBrowserModel implements IDisposable {
     if (path !== '') {
       path = normalizePath(this._model.path, path);
     }
+    let changed = false;
+    if (path !== this.path) {
+      this._selection = {};
```
Comment (sccolbert) [inline src/model.ts]: `Object.create(null);`

**Comment by sccolbert** [comment]
:shipit: after fixing comments

File: src/model.ts (hunk 6 above)
Comment (blink1073) [inline src/model.ts:169]: Nevermind, it worked.

**Pull request merged**